azurelinuxagent/ga/exthandlers.py

```diff
@@ -1396,3 +1437,158 @@ def is_update_with_install(self):
         return update_mode.lower() == "updatewithinstall"
+
+
+class HandlerConfiguration(object):
+    @staticmethod
+    def send_handler_configuration_event(name=AGENT_NAME, message="", operation=WALAEventOperation.Unknown,
+                                         is_success=True, log_event=True):
+        add_event(name=name, message=message, op=operation, is_success=is_success, log_event=log_event)
+
+    def __init__(self, data):
+        if data is None:
+            raise ExtensionConfigurationError('Malformed handler configuration file.')
+        if "handlerConfiguration" not in data:
+            raise ExtensionConfigurationError('Malformed handler configuration file.')
+        if "linux" not in data['handlerConfiguration']:
+            raise ExtensionConfigurationError('No linux configurations present in HandlerConfiguration')
+
+        self.data = data
+
+    def get_name(self):
+        return self.data["name"]
+
+    def get_version(self):
+        return self.data["version"]
+
+    def get_linux_configurations(self):
+        return self.data['handlerConfiguration']['linux']
+
+    def get_resource_configurations(self):
+        if "resources" in self.get_linux_configurations() and (
+                "cpu" in self.get_linux_configurations()["resources"] or
+                "memory" in self.get_linux_configurations()["resources"]):
+            return self.get_linux_configurations()["resources"]
+        return None
+
+    def get_cpu_limits_for_extension(self):
+        resource_config = self.get_resource_configurations()
+
+        try:
+            if resource_config and "cpu" in resource_config:
+                return CpuLimits(resource_config["cpu"])
+        except ExtensionConfigurationError as e:
+            logger.warn(ustr(e))
+            HandlerConfiguration.send_handler_configuration_event(message=ustr(e), is_success=False, log_event=False,
+                                                                  operation=WALAEventOperation.HandlerConfiguration)
+        return None
+
+    def get_memory_limits_for_extension(self):
+        resource_config = self.get_resource_configurations()
+
+        try:
+            if resource_config and "memory" in resource_config:
+                return MemoryLimits(resource_config["memory"])
+        except ExtensionConfigurationError as e:
+            logger.warn(ustr(e))
+            HandlerConfiguration.send_handler_configuration_event(message="{0}".format(ustr(e)), is_success=False,
+                                                                  log_event=False,
+                                                                  operation=WALAEventOperation.HandlerConfiguration)
+
+        return None
+
+
+class CpuLimits(object):
+    def __init__(self, cpu_node):
+        self.cpu_limits = []
+        self.cores = []
+
+        for property in cpu_node:
+            # integers and > 0.
+            if "cores" not in property or "limit_percentage" not in property:
```

Comment:
we should do all the validation of the configuration file when we load it

Reply:
I do 4-level validation at the time of loading up itself, but doing a very nested level validation would be very time-consuming.

Eg: In the following, I validate the following keys for absence and structure. 
level | key
1 - `name`, `version`, `handlerConfiguration`, 
2 - `handlerConfiguration.linux`, 
3 - `handlerConfiguration.linux.resources`, 
4 - `handlerConfiguration.linux.resources.{cpu|memory}`
```
{
      "name": "ExampleHandlerLinux",
      "version": 1.0,
      "handlerConfiguration": {
        "linux": {
          "resources": {
            "cpu": [
              {
                "cores": 2,
                "limit_percentage": 25
              },
              {
                "cores": 8,
                "limit_percentage": 20
              },
              {
                "cores": -1,
                "limit_percentage": 15
              }
            ],
            "memory": {
              "max_limit_percentage": 20,
              "max_limit_MBs": 1000,
              "memory_pressure_warning": "low",
              "memory_oom_kill": "enabled"
            }
          }
        },
        "windows": {}
      }
    }
```

We can discuss the viability of doing a rigid check of the whole file, but I am erroring on the side of publishers knowing the structure. Let me know your thoughts.

Reply:
i would add more validation, including typos in the values (e.g. "cores" should be an integer)

the idea is that extension publishers should be able to find any errors/typos in their initial testing, and not after the extension has been deployed.

can you look if DataContract facilitates validation? (I have not read the code).

i would not expect validation to take more than tenths or seconds at worst, do you think validation time is a concern? thanks

Reply:
> the idea is that extension publishers should be able to find any errors/typos in their initial testing, and not after the extension has been deployed.
Just for my clarification - the extension will still be deployed, as we don't want to deny users using the extension only because cpu/memory values were not set correctly? This might not be true in the case of capabilities, as we don't want to setup something with more privilege (which can then be actively done in `ExtensionCapabilitiesConfiguration`-like class).

and the extension, while being deployed is going to hit the validation points, where the telemetry would capture the validation issues, if any. This would be exercised by the `CGroupsTelemetry.track_extension(self.ext_handler.name, cg, handler_configuration.get_resource_configurations())` call after the launch of the extension is done.

> can you look if DataContract facilitates validation? (I have not read the code).
I don't think so. I only a simple fetch of data done by the `get_properties(obj)` method using the DataContract. Its a verly loosely specified.
```
def get_properties(obj):
    if isinstance(obj, DataContract):
        data = {}
        props = vars(obj)
        for prob_name, prob in list(props.items()):
            data[prob_name] = get_properties(prob)
        return data
    elif isinstance(obj, DataContractList):
        data = []
        for item in obj:
            item_data = get_properties(item)
            data.append(item_data)
        return data
    else:
        return obj

class DataContract(object):
    pass
```

> i would not expect validation to take more than tenths or seconds at worst, do you think validation time is a concern? thanks
It isn't not much, but I think `handlerConfig` not be something that all extensions would be signing up for, and would rely more on the default limits, and all the work we are doing on validation is mostly not needed.

I am okay to do the change - let me check what all things are required for the changes.

Reply:
@narrieta - This is now done. The HandlerConfig is validated at load time. After going through the usecases for tests, it made sense to load and report the issues immediately. Thanks.